server/test/services/apikey/ApiKeyServiceTest.java

```diff
@@ -0,0 +1,177 @@
+
+    assertThat(apiKeyCreationResult.getForm().error("keyName").get().message())
+        .isEqualTo("Key name cannot be blank.");
+  }
+
+  @Test
+  public void createApiKey_missingExpiration_reportsError() throws ProgramNotFoundException {
+    DynamicForm form =
+        buildForm(
+            ImmutableMap.of(
+                "keyName", "test key",
+                "subnet", "0.0.0.1/32"));
+
+    ApiKeyCreationResult apiKeyCreationResult = apiKeyService.createApiKey(form, adminProfile);
+
+    assertThat(apiKeyCreationResult.getForm().error("expiration").get().message())
+        .isEqualTo("Expiration cannot be blank.");
+  }
+
+  @Test
+  public void createApiKey_malformedExpiration_reportsError() throws ProgramNotFoundException {
+    DynamicForm form =
+        buildForm(
+            ImmutableMap.of(
+                "keyName", "test key",
+                "expiration", "01-2020-30",
+                "subnet", "0.0.0.1/32"));
+
+    ApiKeyCreationResult apiKeyCreationResult = apiKeyService.createApiKey(form, adminProfile);
+
+    assertThat(apiKeyCreationResult.getForm().error("expiration").get().message())
+        .isEqualTo("Expiration must be in the form YYYY-MM-DD.");
+  }
+
+  @Test
+  public void createApiKey_missingSubnet_reportsError() throws ProgramNotFoundException {
+    DynamicForm form =
+        buildForm(
+            ImmutableMap.of(
+                "keyName", "test key",
+                "expiration", "2020-01-30"));
+
+    ApiKeyCreationResult apiKeyCreationResult = apiKeyService.createApiKey(form, adminProfile);
+
+    assertThat(apiKeyCreationResult.getForm().error("subnet").get().message())
+        .isEqualTo("Subnet cannot be blank.");
+  }
+
+  @Test
+  public void createApiKey_malformedSubnet_reportsError() throws ProgramNotFoundException {
+    DynamicForm form =
+        buildForm(
+            ImmutableMap.of(
+                "keyName", "test key",
+                "expiration", "2020-01-30",
+                "subnet", "0.1.1.1"));
+
+    ApiKeyCreationResult apiKeyCreationResult = apiKeyService.createApiKey(form, adminProfile);
+
+    assertThat(apiKeyCreationResult.getForm().error("subnet").get().message())
+        .isEqualTo("Subnet must be in CIDR notation.");
+  }
+
+  @Test
+  public void createApiKey_grantedProgramSlugNotFound_raisesProgramNotFound()
+      throws ProgramNotFoundException {
+    DynamicForm form =
+        buildForm(
+            ImmutableMap.of(
+                "keyName", "test key",
+                "expiration", "2020-01-30",
+                "subnet", "0.1.1.1",
+                "grant-program-read[test-program]", "true"));
+
+    assertThrows(
+        ProgramNotFoundException.class, () -> apiKeyService.createApiKey(form, adminProfile));
+  }
+
+  private DynamicForm buildForm(ImmutableMap<String, String> formContents) {
+    return formFactory.form().bindFromRequest(fakeRequest().bodyForm(formContents).build());
+  }
+}
```

Comment:
Add a test for "ApiKey creator must have authority_id"?

Reply:
done